### PR TITLE
More work on deserialization

### PR DESCRIPTION
**Pull request by pguillory**
- Filled in some missing functionality, like deserializing lists of sets or maps.
- Added a little documentation. It doesn't communicate the idea well at the
  moment.
- Re-organized a bit.
- Cleaned up whitespace and commented out code.

### CODE REVIEW

File: test/generator/binary_protocol_test.exs [hunk 1]
```diff
@@ -1,5 +1,6 @@
 defmodule Thrift.Generator.BinaryProtocolTest do
-  use ExUnit.Case, async: true
+  use ThriftTestCase, cleanup: false
+  # use ExUnit.Case, async: true
```
Comment (jparise) [inline test/generator/binary_protocol_test.exs]: Drop this?

File: lib/generator/binary_protocol.ex [hunk 2]
```diff
@@ -1,57 +1,102 @@
 defmodule Thrift.Generator.Models.BinaryProtocol do
```
Comment (scohen) [inline lib/generator/binary_protocol.ex]: This module needs to move out of generator and into Thrift.Protocols.Binary

File: lib/generator/binary_protocol.ex [hunk 3]
```diff
@@ -1,57 +1,102 @@
...
     |> Enum.map(&field_deserializer(&1.type, &1, :deserialize, file_group))
     |> merge_blocks
 
     quote do
-      # def serialize(unquote(struct_matcher)) do
-      #   unquote([field_serializers, <<0>>] |> merge_binaries |> merge_binaries)
-      # end
+      # defp bool_to_int(nil), do: 0
+      # defp bool_to_int(false), do: 0
+      # defp bool_to_int(_), do: 1
```
Comment (scohen) [inline lib/generator/binary_protocol.ex]: remove these commented-out lines?

File: lib/generator/binary_protocol.ex [hunk 4]
```diff
@@ -1,57 +1,102 @@
...
+
+      def serialize2(unquote(struct_matcher)) do
```
Comment (scohen) [inline lib/generator/binary_protocol.ex]: ;-/

File: test/generator/binary_protocol_test.exs (hunk 1 above)
Comment (pguillory) [inline test/generator/binary_protocol_test.exs]: Forgot I'd intended to convert this to use ThriftTestCase. Done now.

File: lib/generator/binary_protocol.ex (hunk 4 above)
Comment (pguillory) [inline lib/generator/binary_protocol.ex]: I uncommented this stuff to do some more benchmarking. If it's faster, we ought to use it.

Comment (scohen) [inline lib/generator/binary_protocol.ex]: I'm just commenting on the name.

Comment (scohen) [inline lib/generator/binary_protocol.ex]: How fast is it?

**Pull request merged**